### ISSUE TITLE
control_flow_test is failing on vulkan backend

### ISSUE THREAD

**Issue by hanhanW** [google/iree#2526]
The test started failing after #2509, and #2522 disabled the test: https://github.com/google/iree/blob/main/integrations/tensorflow/e2e/control_flow_test.py

Seems that there is a runtime error issue: `RuntimeError: Error invoking function: 8`. Not sure if this is pyiree issue or vulkan runtime issue or others.

This issue is for tracking the failing test.

**Comment by ScottTodd** [comment]
@phoenix-meadowlark and I looked into this one briefly yesterday

8 is `IREE_STATUS_RESOURCE_EXHAUSTED`: https://github.com/google/iree/blob/50d264b75f7e68d2af5c698a94daa70f586c5ee7/iree/base/api.h#L219

The error that was hit was here (from logs that aren't public >_>):
https://github.com/google/iree/blob/50d264b75f7e68d2af5c698a94daa70f586c5ee7/iree/hal/vulkan/timepoint_util.cc#L68-L75
That will happen with SwiftShader or this flag: https://github.com/google/iree/blob/50d264b75f7e68d2af5c698a94daa70f586c5ee7/iree/hal/vulkan/vulkan_device.cc#L45-L46

A workaround is to increase this value: https://github.com/google/iree/blob/50d264b75f7e68d2af5c698a94daa70f586c5ee7/iree/hal/vulkan/timepoint_util.h#L118-L120

Or we can change program behavior to require less outstanding fences (@antiagainst, ideas?)

**Comment by benvanik** [comment]
That would indicate a bug - right now we literally only use the semaphores to wait and never overlap anything, so there should only ever be a single live fence at a time.

**Comment by benvanik** [comment]
(semaphore created, submitted, and waited on here, then destroyed when the vm::ref goes out of scope) https://github.com/google/iree/blob/4c0a53e955c44480e2387a9d756a33f66830a394/iree/modules/hal/hal_module.cc#L141-L165

**Comment by antiagainst** [comment]
Okay I'll take a look. Thanks for reporting! :)
